integration/client-data-test.ts

```diff
@@ -0,0 +1,357 @@
+    `,
+  };
+}
+
+test.describe("Client Data", () => {
+  let appFixture: AppFixture;
+
+  test.afterAll(() => {
+    appFixture.close();
+  });
+
+  test.describe("Initial Hydration", () => {
+    test("no client loaders or fallbacks", async ({ page }) => {
+      appFixture = await createAppFixture(
+        await createFixture({
+          files: getFiles({ parentAdditions: "", childAdditions: "" }),
+        })
+      );
+      let app = new PlaywrightFixture(appFixture, page);
+
+      // Full SSR - normal Remix behavior due to lack of clientLoader
+      await app.goto("/parent/child");
+      let html = await app.getHtml("main");
+      expect(html).toMatch("Parent Component");
+      expect(html).toMatch("Parent Server Loader");
+      expect(html).toMatch("Child Component");
+      expect(html).toMatch("Child Server Loader");
+    });
+
+    test("parent.clientLoader/child.clientLoader", async ({ page }) => {
+      appFixture = await createAppFixture(
+        await createFixture({
+          files: getFiles({
+            parentAdditions: js`
+            export async function clientLoader() {
+              await new Promise(r => setTimeout(r, 100));
+              return { message: "Parent Client Loader" };
+            }
+          `,
+            childAdditions: js`
+            export async function clientLoader() {
+              await new Promise(r => setTimeout(r, 100));
+              return { message: "Child Client Loader" };
+            }
+          `,
+          }),
+        })
+      );
+      let app = new PlaywrightFixture(appFixture, page);
+
+      // Full SSR - normal Remix behavior due to lack of Fallback components
+      await app.goto("/parent/child");
+      let html = await app.getHtml("main");
+      expect(html).toMatch("Parent Component");
+      expect(html).toMatch("Parent Server Loader");
+      expect(html).toMatch("Child Component");
+      expect(html).toMatch("Child Server Loader");
+    });
+
+    test("parent.clientLoader/parent.Fallback/child.clientLoader", async ({
+      page,
+    }) => {
+      appFixture = await createAppFixture(
+        await createFixture({
+          files: getFiles({
+            parentAdditions: js`
+            export async function clientLoader() {
+              await new Promise(r => setTimeout(r, 100));
+              return { message: "Parent Client Loader" };
+            }
+            export function Fallback() {
+              return <p>Parent Fallback</p>
+            }
+          `,
+            childAdditions: js`
+            export async function clientLoader() {
+              await new Promise(r => setTimeout(r, 100));
+              return { message: "Child Client Loader" };
+            }
+          `,
+          }),
+        })
+      );
+      let app = new PlaywrightFixture(appFixture, page);
+
+      // Renders parent fallback on initial render and calls parent clientLoader
+      // Does not call child clientLoader due to lack of Fallback
```

Comment:
I think a parent route fallback should enable child route client loaders to run, because the fallback can wait for any nested route client loader too.

Similar to how Suspense works, if I have:
```ts
<Suspense fallback={<Fallback />}>
  <ParentRoute>
    <ChildRoute />
  </ParentRoute>
</Suspense>
```
The Suspense will render the fallback until both ParentRoute and ChildRoute renders.

But thanks to <Outlet /> Remix can enable the ChildRoute to render inside the ParentRoute fallback component too, similar to this:

```ts
let child = (
  <Suspense fallback={<ChildFallback />}>
    <ChildRoute />
  </Suspense>
)

<Suspense fallback={<ParentFallback>{child}</ParentFallback>}>
  <ParentRoute>
    {child}
  </ParentRoute>
</Suspense>
```

Reply:
yeah I've gone back and forth on this - planning to talk it over with @mjackson and @ryanflorence.  On one hand I really like the fact that the current implementation allows me to answer the "does this run on hydration" question from the same file.  But what you are proposing is more in line with Suspense concepts - but it makes it a bit harder to track down since now you need to walk up the route hierarchy yourself.

I also don't want to introduce uneccesary calls - for example if a child `clientLoader` doesn't want to run on hydration but some parent does - then we'd call the child `clientLoader` for no reason.  But we are planning to provide a way to know in the `clientLoader` if it's the hydration call or not (currently via a custom header) - so you could short circuit manually:

```js
function clientLoader({ request, serverLoader }) {
  if (isHydration(request)) {
    return serverLoader();
  }
  // logic for subsequent SPA navigations
}
```